moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java

```diff
@@ -145,82 +139,58 @@ private void recreateSessionPool() {
+                terminateSession(previous);
             }
         } else {
-            postConnectAction = reopenExistingSession(msg, clientId, oldSession, newSession, username);
+            postConnectAction = reopenExistingSession(clientId, oldSession, username);
         }
+
         return postConnectAction;
     }
 
-    private SessionCreationResult reopenExistingSession(MqttConnectMessage msg, String clientId,
-                                                        Session oldSession, Session newSession, String username) {
-        final boolean newIsClean = msg.variableHeader().isCleanSession();
+    private SessionCreationResult reopenExistingSession(String clientId, Session oldSession, String username) {
         final SessionCreationResult creationResult;
         if (oldSession.disconnected()) {
-            if (newIsClean) {
-                boolean result = oldSession.assignState(SessionStatus.DISCONNECTED, SessionStatus.CONNECTING);
-                if (!result) {
-                    throw new SessionCorruptedException("old session was already changed state");
-                }
-
-                // case 2
-                // publish new session
-                dropQueuesForClient(clientId);
-                unsubscribe(oldSession);
-                copySessionConfig(msg, oldSession);
-
-                LOG.trace("case 2, oldSession with same CId {} disconnected", clientId);
-                creationResult = new SessionCreationResult(oldSession, CreationModeEnum.CREATED_CLEAN_NEW, true);
-            } else {
-                final boolean connecting = oldSession.assignState(SessionStatus.DISCONNECTED, SessionStatus.CONNECTING);
-                if (!connecting) {
-                    throw new SessionCorruptedException("old session moved in connected state by other thread");
-                }
-                // case 3
-                reactivateSubscriptions(oldSession, username);
-
-                LOG.trace("case 3, oldSession with same CId {} disconnected", clientId);
-                creationResult = new SessionCreationResult(oldSession, CreationModeEnum.REOPEN_EXISTING, true);
+            final boolean connecting = oldSession.assignState(SessionStatus.DISCONNECTED, SessionStatus.CONNECTING);
+            if (!connecting) {
+                throw new SessionCorruptedException("old session moved in connected state by other thread");
             }
-        } else {
-            // case 4
-            LOG.trace("case 4, oldSession with same CId {} still connected, force to close", clientId);
-            oldSession.closeImmediately();
-            //remove(clientId);
-            creationResult = new SessionCreationResult(newSession, CreationModeEnum.DROP_EXISTING, true);
-        }
+            // case 2
+            reactivateSubscriptions(oldSession, username);
 
-        if (creationResult.mode == CreationModeEnum.DROP_EXISTING) {
```

Comment:
This section still seems relevant to me.
If the old session is not disconnected and new session is not clean, we have to drop the existing one. 

Is there some place else you are handling this case?

Reply:
isn't that the case 3 at line 179 in the new code? oldsession would not be null and it is not clean for the new session, so it does 
```
            postConnectAction = reopenExistingSession(clientId, oldSession, username);
```

if the oldSession is not disconnected, it goes to the else branch here

Reply:
Yep, Rob is right. This is only dealing with persistent sessions at this point. So the Session is being resumed, regardless of whether the old client is connected. Old connection is disconnected at line 180